src/sync/sync_session.cpp

```diff
@@ -473,18 +474,22 @@ void SyncSession::handle_progress_update(uint64_t downloaded, uint64_t downloada
     }
 }
 
-void SyncSession::NotifierPackage::update(const Progress& current_progress)
+void SyncSession::NotifierPackage::update(const Progress& current_progress, bool data_is_fresh)
 {
-    if (is_streaming || captured_transferrable)
+    if (is_streaming || captured_transferrable || !data_is_fresh)
         return;
 
     captured_transferrable = direction == NotifierType::download ? current_progress.downloadable
                                                                  : current_progress.uploadable;
 }
 
-std::function<void()> SyncSession::NotifierPackage::create_invocation(const Progress& current_progress, bool& is_expired) const
+std::function<void()> SyncSession::NotifierPackage::create_invocation(const Progress& current_progress,
+                                                                      bool& is_expired) const
 {
-    REALM_ASSERT(is_streaming || captured_transferrable);
+    // It's possible for a non-streaming notifier to not yet have fresh transferrable bytes data.
+    // In that case, we don't call it at all.
```

Comment:
The idea is that we don't call the callback until sync reports that it has numbers that came from the server recently ('fresh', not 'stale').

The problem was that people were opening synced Realms and immediately registering notifier blocks. Sync would happily report 0 out of 0 bytes before it actually got any info from the server, so the notifier blocks would say "your download has finished". Then the server would report fresh data and sync would report (e.g.) 0 out of 20 bytes, but by then it was too late.